search/src/dockit-search.ts

```diff
@@ -0,0 +1,260 @@
+import { LitElement, css, html } from 'lit';
+import { unsafeHTML } from 'lit/directives/unsafe-html.js';
+
+const beforeHitMatch = 20;
+const afterHitMatch = beforeHitMatch;
+const markChar = '\0';
+
+export class SearchBox extends LitElement {
+  search = null;
+  hits = null;
+
+  constructor() {
+    super();
+    addEventListener('keydown', (e) => {
+      if ((e.ctrlKey || e.metaKey) && e.key === 'k') {
+        e.preventDefault();
+        this.updateHits();
+      } else if (e.key === 'Escape') this.hideHits();
+    });
+  }
+
+  render() {
+    return html`<div id="search-box">
+      <form action="" role="search" novalidate="">
+        <label for="search-input" aria-label="Search">
+          <input
+            aria-autocomplete="list"
+            id="search-input"
+            autocomplete="off"
+            autocorrect="off"
+            autocapitalize="off"
+            spellcheck="false"
+            maxlength="64"
+            type="search"
+            enterkeyhint="go"
+            placeholder=${this.placeHolder}
+            @input=${(e: { target: { value: string } }) =>
+              this.updateHits(e.target.value)}
+            @focus=${(e: { target: { value: string } }) =>
+              this.updateHits(e.target.value)}
+          />
+        </label>
+      </form>
+      <ul id="search-hits">
+        ${this.hits === null
+          ? html`<li><i>Type for searching...</i></li>`
+          : this.hits.length === 0
+          ? html`<li><i>No match found</i></li>`
+          : this.hits.map(
+              (hit) =>
+                html`<li>
+                  <a href="${hit.id}" @click=${this.hideHits}>
+                    ${unsafeHTML(this.hitHTML(hit))}
+                  </a>
+                </li>`
+            )}
+      </ul>
+      <div id="search-overlay" @click=${this.hideHits}></div>
+    </div>`;
+  }
+
+  hitHTML(hit) {
```

Comment:
most of the methods of this class should be private, and can you please also type them fully?

Reply:
done